// @vygruppen/spor-react/dist/CountryCodeSelect-2DASZ3PQ.mjs
import { createTexts, useTranslation, InfoSelect, Item } from './chunk-MDMLROK2.mjs';
import React from 'react';
import { getSupportedCallingCodes } from 'awesome-phonenumber';

var prioritizedCountryCodes = [
  { key: "+47", value: "+47" },
  { key: "+46", value: "+46" },
  { key: "+45", value: "+45" }
];
var sortedCallingCodes = getSupportedCallingCodes().sort((a, b) => Number(a) - Number(b)).map((code) => ({
  key: `+${code}`,
  value: `+${code}`
})).filter(
  (code) => !prioritizedCountryCodes.some((pCode) => pCode.key === code.key)
);
var callingCodes = [...prioritizedCountryCodes, ...sortedCallingCodes];
var CountryCodeSelect = (props) => {
  const { t } = useTranslation();
  return /* @__PURE__ */ React.createElement(
    InfoSelect,
    {
      label: t(texts.countryCode),
      isLabelSrOnly: true,
      items: callingCodes,
      ...props
    },
    (item) => /* @__PURE__ */ React.createElement(Item, { key: item.key }, item.key)
  );
};
var CountryCodeSelect_default = CountryCodeSelect;
var texts = createTexts({
  countryCode: {
    nb: "Landkode",
    nn: "Landskode",
    en: "Country code",
    sv: "Landskod"
  }
});

export { CountryCodeSelect, CountryCodeSelect_default as default };
